### PR TITLE
Replace Apache HttpClient 3.x with Jetty

### PR DESCRIPTION
**Pull request by svilenvul**
Apache HttpClient is replaced in 3 smarthome bundles.

Added RetryListener (which should be a replacement for the
DefaultHttpMethodRetryHandler in Apache 3.x library) and method in
HttpUtil class, that should make preemptive authentication easier.

After this change Apache HttpClient 3.x. could be removed from smarthome
target platform.

Discussion: #1689 

Signed-off-by: Svilen Valkanov svilen.valkanov@musala.com

P.S.
Forgot to mention #682 as well.

### CODE REVIEW

**Comment by maggu2810** [comment]
Can you change the Signed-off-by line to the commonly used one?

```
Signed-off-by: Joe Somebody <somebody@someplace.net>
```

Then the IP-Validation should not fail if you have a valid CLA

File: extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -7,20 +7,22 @@ Bundle-Version: 0.9.0.qualifier
...
  org.eclipse.smarthome.core.thing,
  org.eclipse.smarthome.core.thing.binding,
  org.eclipse.smarthome.core.thing.binding.builder,
  org.eclipse.smarthome.core.thing.type,
  org.eclipse.smarthome.core.types,
- org.eclipse.smarthome.config.discovery,
  org.jupnp.model.meta,
- org.slf4j
+ org.slf4j,
+ org.eclipse.smarthome.io.net.http
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/META-INF/MANIFEST.MF]: Can you keep the alphabetical order (for manifest first it is perhaps easier to manage if it is ordered).

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package org.eclipse.smarthome.io.net.http;
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java]: Missing header

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 3]
```diff
@@ -152,63 +156,71 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
+        Map<String, String> credentials = extractCredentials(url);
         if (credentials != null) {
-            client.getParams().setAuthenticationPreemptive(true);
-            client.getState().setCredentials(AuthScope.ANY, credentials);
+            Authentication.Result result = createBasicAuthenticationResult(url, 0, credentials.get(USERNAME_KEY),
+                    credentials.get(PASSWORD_KEY), false);
+            client.getAuthenticationStore().addAuthenticationResult(result);
         }
 
         Logger logger = LoggerFactory.getLogger(HttpUtil.class);
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: You have already introduced a static logger, so you can use LoggerFactory.getLogger in your static functions and drop the static member or you can use the static member and drop the getLogger calls.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 4]
```diff
@@ -152,63 +156,71 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
         if (logger.isDebugEnabled()) {
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: It has not been introduced by you, but we normally rely on the fact that logger.debug is inexpensive if debugging severity is not used.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 5]
```diff
@@ -152,63 +156,71 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
-            try {
-                logger.debug("About to execute '" + method.getURI().toString() + "'");
-            } catch (URIException e) {
-                logger.debug(e.getMessage());
-            }
+            logger.debug("About to execute '" + request.getURI().toString() + "'");
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: This has been taken from the previous version, but we should not build strings in the logger argument and let's build the string by the logger.

`logger.debug("About to execute {}", request.getURI());`

If request.getURI() is not trivial, we can keep the isDebugEnabled call above.

File: extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/src/main/java/org/eclipse/smarthome/binding/fsinternetradio/internal/radio/FrontierSiliconRadioConnection.java [hunk 6]
```diff
@@ -185,20 +184,38 @@ public FrontierSiliconRadioApiResult doRequest(String requestString, String para
...
         }
         isLoggedIn = false; // 3 tries failed. log in again next time, maybe our session went invalid (radio restarted?)
         return null;
     }
+
+    private void startHttpClient(HttpClient client) {
+        if (httpClient == null) {
+            httpClient = new HttpClient();
+        }
+        if (!client.isStarted()) {
```
Comment (maggu2810) [inline extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/src/main/java/org/eclipse/smarthome/binding/fsinternetradio/internal/radio/FrontierSiliconRadioConnection.java]: Do you mix 'httpClient' and `client` by intention this way?

**Comment by maggu2810** [comment]
Thanks @svilenvul! I have added some comments.

**Comment by svilenvul** [comment]
Thanks @maggu2810 
I have a valid CLA for svilen.valkanov@musala.com, but the IP Validation failed again.
I will take a look how to fix that in my next commit.

**Comment by kaikreuzer** [comment]
Try adding the brackets around your email address.

**Comment by maggu2810** [comment]
For me okay if all stuff is still working.
@kaikreuzer Can we give it a try?

**Comment by kaikreuzer** [comment]
> if all stuff is still working.

I'd like to do some testing first, didn't yet find the time for it.
But I very much welcome the PR, many thanks already, @svilenvul!

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 7]
```diff
@@ -59,6 +60,11 @@
      *
      * @return the response body or <code>NULL</code> when the request went wrong
      */
+
+    private static final String USERNAME_KEY = "user";
+    private static final String PASSWORD_KEY = "pass";
+    private static final int DEFAULT_NUMBER_OF_RETRIES = 3;
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: I see that you try to keep the old functionality in place, but I actually wonder, if an automatic retry in this util class makes sense - I never saw this being useful anywhere yet and retry handling might better be left to the caller. What do others think?

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 8]
```diff
@@ -152,63 +157,70 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
 
         HttpClient client = new HttpClient();
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: Reading the [docs](http://download.eclipse.org/jetty/stable-9/apidocs/org/eclipse/jetty/client/HttpClient.html), it says "HttpClient acts as a central configuration point " and "HttpClient transparently pools connections to servers".
I know that creating a new instance is usually an expensive operation and we are doing it here for every single request. I am wondering, if we can reuse the http client for all our calls? Or is there a good reason to have a separate instance for each?

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 9]
```diff
@@ -152,63 +157,70 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
 
+        try {
+            client.start();
+        } catch (Exception e1) {
+            logger.warn("HttpClient can not be stared !", e1);
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: typo: started

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 10]
```diff
@@ -152,63 +157,70 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
         }
 
-        HttpMethod method = HttpUtil.createHttpMethod(httpMethod, url);
-        method.getParams().setSoTimeout(timeout);
-        method.getParams().setParameter(HttpMethodParams.RETRY_HANDLER, new DefaultHttpMethodRetryHandler(3, false));
+        HttpMethod method = HttpUtil.createHttpMethod(httpMethod);
+
+        Request request = client.newRequest(url).method(method).timeout(timeout, TimeUnit.MILLISECONDS);
+        RetryListener listener = new RetryListener(request, DEFAULT_NUMBER_OF_RETRIES);
+        request.onComplete(listener);
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: You retry logic is not working. Your listener will be informed, but your logic in this method will continue and simply return null on a first unsuccessful request. So even if requests 2, 3 or 4 are successful, the caller won't get the result.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 11]
```diff
@@ -294,28 +309,73 @@ protected static Credentials extractCredentials(String url) {
...
+     * When this result is added to the {@link AuthenticationStore},
+     * the {@link HttpClient} will send authentication header preemptively.
+     *
+     * @param host - the hostname
+     * @param port - the port, if set to not positive number it will be ignored
+     * @param user - the name of the user that is requiring the resource
+     * @param password - the user password
+     * @param isProxyAuthentication - parameter that specifies, if the result will be used for proxy authenticaiton
+     */
+    public static Authentication.Result createBasicAuthenticationResult(String host, int port, String user,
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: BasicAuth support seems to be built-in already, see https://www.eclipse.org/jetty/documentation/current/http-client-other.html#http-client-authentication.

So I assume you won't need this method?

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java (hunk 8 above)
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: Correct, normally it should be enough to create one client in constructor, start it on initialize and stop it on dispose.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    @Override
+    public void onComplete(Result result) {
+
+        if (retries > 0 && result.isFailed()) {
+
+            logger.debug("Request failed with Exception.", result.getFailure());
+            --retries;
+            try {
+                logger.debug("Trying to resend request");
+                request.send();
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java]: This does not deal in any way with the response or its status.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is an implementation of {@link Response.CompleteListener}, which retries an HTTP Request, if it has failed.
+ *
+ * @author Svilen Valkanov - initial contribution
+ */
+
+public class RetryListener implements Response.CompleteListener {
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java]: This class imho does not need to be public (and be in an exported package).
As discussed on the HttpUtil, I would tend to say we can remove it completely.

File: bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/proxy/ProxyServlet.java [hunk 14]
```diff
@@ -158,31 +164,57 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
         HttpClient httpClient = new HttpClient();
```
Comment (kaikreuzer) [inline bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/proxy/ProxyServlet.java]: same here as in HttpUtil: Let's reuse the client instance within the servlet.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java (hunk 11 above)
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: Besides this, you are making Authentication.Result part of the API of HttpUtil, which is not good (it would make a refactoring like the one in this PR impossible).

File: extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/src/main/java/org/eclipse/smarthome/binding/fsinternetradio/internal/radio/FrontierSiliconRadioConnection.java [hunk 15]
```diff
@@ -9,23 +9,23 @@
...
+import org.eclipse.smarthome.io.net.http.RetryListener;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * This class holds the http-connection and session information for controlling the radio.
  *
  * @author Rainer Ostendorf
- * @author Patrick Koenemann
+ * @author Patrick Koeneman
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/src/main/java/org/eclipse/smarthome/binding/fsinternetradio/internal/radio/FrontierSiliconRadioConnection.java]: why did you remove an "n" on his name?

File: extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/META-INF/MANIFEST.MF [hunk 16]
```diff
@@ -7,18 +7,20 @@ Bundle-Version: 0.9.0.qualifier
...
  org.eclipse.smarthome.config.core,
+ org.eclipse.smarthome.config.discovery,
  org.eclipse.smarthome.core.library.types,
  org.eclipse.smarthome.core.thing,
  org.eclipse.smarthome.core.thing.binding,
  org.eclipse.smarthome.core.thing.binding.builder,
  org.eclipse.smarthome.core.thing.type,
  org.eclipse.smarthome.core.types,
- org.eclipse.smarthome.config.discovery,
+ org.eclipse.smarthome.io.net.http,
```
Comment (kaikreuzer) [inline extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/META-INF/MANIFEST.MF]: If code is using the Jetty client directly, it should not also use io.net.http - it should imho be an either/or.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java (hunk 11 above)
Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: There is support for BasicAuth, but I could not find any inbuilt support for preemptive authentication
>A client MAY preemptively send the corresponding Authorization header with requests for resources in
   that space **without receipt of another challenge from the server.**

This is from [HTTP Authentication: Basic and Digest Access Authentication chapter chapter 2  Basic Authentication Scheme](https://www.ietf.org/rfc/rfc2617.txt).

The current implementation was using ```client.setAuthenticationPreemptive(true)```, this is why I added this method as a replacement.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java (hunk 10 above)
Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: I will try to write a simple test case for this listener, if the retry functionality is needed at all.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 17]
```diff
@@ -152,63 +157,70 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
                 logger.trace(responseBody);
             }
 
             return responseBody;
-        } catch (HttpException he) {
-            logger.error("Fatal protocol violation: {}", he.toString());
-        } catch (IOException ioe) {
-            logger.error("Fatal transport error: {}", ioe.toString());
+        } catch (Exception e) {
+            logger.error("Fatal transport error: {}", e);
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: This clutters the log with long exceptions, if a host is not available. That's why originally, only the message was logged, not the exception.

File: bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/proxy/ProxyServlet.java [hunk 18]
```diff
@@ -51,6 +55,7 @@
  * This servlet also supports data streams, such as a webcam video stream etc.
  *
  * @author Kai Kreuzer - Initial contribution and API
+ * @author Svilen Valkanov - Replaced Apache HttpClient with Jetty
```
Comment (kaikreuzer) [inline bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/proxy/ProxyServlet.java]: The changes in this class break the functionality of the proxy servlet.
You can try this with the demo sitemap of openHAB (with the "Multimedia" menu entry).
This includes an image and a video using the proxy servlet, which results in these urls:
http://localhost:8080/proxy?sitemap=demo.sitemap&widgetId=02030100
http://localhost:8080/proxy?sitemap=demo.sitemap&widgetId=02030101

Both result in exceptions like:
```
2016-07-22 15:15:14.572 [WARN ] [e.jetty.servlet.ServletHandler:628  ] - /proxyjava.io.IOException: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Buffering capacity exceeded
	at org.eclipse.smarthome.ui.internal.proxy.ProxyServlet.doGet(ProxyServlet.java:209)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:60)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Buffering capacity exceeded
	at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118)
	at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:101)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:653)
	at org.eclipse.smarthome.ui.internal.proxy.ProxyServlet.doGet(ProxyServlet.java:199)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:60)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
Caused by: java.lang.IllegalArgumentException: Buffering capacity exceeded
	at org.eclipse.jetty.client.util.BufferingResponseListener.onHeaders(BufferingResponseListener.java:75)
	at org.eclipse.jetty.client.ResponseNotifier.notifyHeaders(ResponseNotifier.java:103)
	at org.eclipse.jetty.client.ResponseNotifier.notifyHeaders(ResponseNotifier.java:95)
	at org.eclipse.jetty.client.HttpReceiver.responseHeaders(HttpReceiver.java:257)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.headerComplete(HttpReceiverOverHTTP.java:230)
	at org.eclipse.jetty.http.HttpParser.parseHeaders(HttpParser.java:1036)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1257)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:156)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:117)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:69)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:89)
```

**Comment by kaikreuzer** [comment]
Done with my tests for the moment. Please let me know if you have any questions!

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java (hunk 13 above)
Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/RetryListener.java]: @kaikreuzer 
I made it public, because I could not find a replacement for [DefaultHttpMethodRetryHandler](https://hc.apache.org/httpclient-3.x/apidocs/org/apache/commons/httpclient/DefaultHttpMethodRetryHandler.html) from Apache HttpClient 3.x.

The handler is used in [fsinternetaudio](https://github.com/eclipse/smarthome/pull/1864/files#diff-e8aface86649dcb555149ba4b6f0b317L81) as well, and I thought it could it could be used from other bundles as well.

When we do not need it here, I guess I can move it to the fsinternetaudio binding (before that I should fix the onComplete() method of course ) and make it private.

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java (hunk 8 above)
Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: @kaikreuzer @maggu2810 
We use the HttpClient to store authentication data in the [AuthenticationStore ](http://download.eclipse.org/jetty/stable-9/apidocs/) as well, so I guess I should call ```clearAuthenticationResults()``` on every new request and it should be ok.

File: extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/src/main/java/org/eclipse/smarthome/binding/fsinternetradio/internal/radio/FrontierSiliconRadioConnection.java (hunk 15 above)
Comment (svilenvul) [inline extensions/binding/org.eclipse.smarthome.binding.fsinternetradio/src/main/java/org/eclipse/smarthome/binding/fsinternetradio/internal/radio/FrontierSiliconRadioConnection.java]: Sorry for that, it was no by intention, I will fix it of course.

File: bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/proxy/ProxyServlet.java (hunk 18 above)
Comment (svilenvul) [inline bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/proxy/ProxyServlet.java]: Great, I know how I can test it now.

**Comment by svilenvul** [comment]
@kaikreuzer 
I am ready with the changes.

I have tested the ProxyServlet with the video (http://localhost:8080/proxy?sitemap=demo.sitemap&widgetId=02030101) and now is ok.

I found two problems there - the default buffer was too small (2 MB) and the request was synchronous.

**Comment by kaikreuzer** [comment]
Thanks @svilenvul!
Regarding the changes in FrontierSiliconRadioConnection: I would like to ask @paphko whether it is really relevant to implement the retry functionality in there or if he only used it because it was there in the Apache HTTP client. If it is not really needed, I would rather vote for removing the RetryListener completely.

**Comment by paphko** [comment]
@kaikreuzer to be honest, I don't know. That part was originally written by Rainer Ostendorf. But I assume that it also works without retry. Unfortunately, I don't have an IDE available at the moment, so I can't test it.

**Comment by kaikreuzer** [comment]
@svilenvul, I have spend some time testing this and found some issues regarding concurrency (you must not close the httpclient at the end of a request, because other requests can potentially be running (using the same client) and authentication (using the Jetty classes indeed failed in my test, so adding an authentication header manually seems to be the better approach, especially as we do not need to clear the auth store all the time again).

I wanted to create a PR against your branch, but this does not seem to be available in any public github repo. You can find my improvements in [this commit](https://github.com/kaikreuzer/smarthome/commit/0a26ee7dc49ab2d86248f6fdbf771a477b449f8e), please feel free to pull it into your branch.

I didn't do any changed to the fsinternetradio code, maybe you can check what might need to be changed in there as well. Given @paphko's feedback, I would suggest that you remove the retry from there as well to make the code simpler. Once you did, maybe you can provide a jar version of it to @paphko, so he can test it with his radio?

**Comment by svilenvul** [comment]
@kaikreuzer 
Thanks for the changes, that you have made.
@paphko 
How can I send you the jar version of the binding ?

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 19]
```diff
@@ -150,65 +148,76 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
+            // clear proxy list from previous requests
+            List<Proxy> proxies = proxyConfig.getProxies();
+            if (!proxies.isEmpty()) {
+                proxies.clear();
             }
+
+            HttpProxy proxy = new HttpProxy(proxyHost, proxyPort);
+            proxies.add(proxy);
+
+            authStore.addAuthentication(new BasicAuthentication(proxy.getURI(), null, proxyUser, proxyPassword));
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: A last question: Did you test the proxy support? When fixing the authentication, I noticed that passing null as a realm results in an NPE during the request. I would expect the same to happen here. Unfortunately, I didn't find any information how one can say "any realm", which is what we want to do here. But null does not seem to be the right value for this.

Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: @kaikreuzer 
No I didn`t test it.
I tried to find a bundle, that is using the proxy support, but I could not find one. I searched through [this list ](https://github.com/eclipse/smarthome/issues/1689#issuecomment-228032703).
I could not find info about "any realm" as well.

Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: Could you [ask on their mailing list](https://wiki.eclipse.org/Jetty/FAQ#How_do_I_ask_a_question.3F) on how proxy support is supposed to work?

Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: I have asked on the mailing list.
I have found that there is an [update ](http://www.eclipse.org/jetty/documentation/current/http-client-authentication.html)in the jetty documentation, since the last time, I have visited it.
It is explained how we can make a preemptive authentication without manually creating an "Authenticaiton" header:
```
AuthenticationStore auth = httpClient.getAuthenticationStore();
URI uri = URI.create("http://domain.com/secure");
auth.addAuthenticationResult(new BasicAuthentication.BasicResult(uri, "username", "password"));
```
The same can be used for  [proxy authentication](http://www.eclipse.org/jetty/documentation/current/http-client-proxy.html#http-client-proxy-authentication).
Maybe it is better to make a new commit with the changes.

Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: I had read this as well, but noticed that BasicAuthentication.BasicResult wasn't available in our version (might be a very recent addition).
In any case, adding the header is imho much better, since it is stateless and we do not have to deal with removing it from the auth store again (and dealing with all associated concurrency issues). For the proxy, which is a global setting in Java, it is in contrast imho no problem to put its credentials in the auth store.

Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: Me:
> Is there any option like in the Apache HttpClient API to set credentials for any realm ?

Answer:
> Not currently, please file an issue.

> As a workaround, you can override BasicAuthentication.matches(...)
with your own logic.

I will go with the workaround then. I have set up a local proxy server and I will test the change. I will update the PR with the changes then.

Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: ok, thanks for the update. And don't forget to file an issue at Jetty as they have requested!

Comment (svilenvul) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: [Issue](https://github.com/eclipse/jetty.project/issues/806) is created.
The [workaround](https://github.com/eclipse/smarthome/pull/1864/commits/1208e23ba60212f6189944aa21f4b0e9aab8c435) is ready as well and has been tested with the AstroDiscoveryService (that is using the HttpUtil class).

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java [hunk 20]
```diff
@@ -150,65 +148,87 @@ public static String executeUrl(String httpMethod, String url, Properties httpHe
...
-            if (StringUtils.isNotBlank(proxyUser)) {
-                client.getState().setProxyCredentials(AuthScope.ANY,
-                        new UsernamePasswordCredentials(proxyUser, proxyPassword));
+            AuthenticationStore authStore = client.getAuthenticationStore();
+            ProxyConfiguration proxyConfig = client.getProxyConfiguration();
+
+            // clear proxy list from previous requests
+            List<Proxy> proxies = proxyConfig.getProxies();
+            if (!proxies.isEmpty()) {
+                proxies.clear();
```
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: As we are using the same store for possibly concurrent requests, you must not clear all proxies here.
Instead you should remove the proxy that you have added for this request after the request is done - avoid any side effects on concurrently running requests!

File: bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java (hunk 19 above)
Comment (kaikreuzer) [inline bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpUtil.java]: Cool, thanks - looks good!

**Comment by kaikreuzer** [comment]
> How can I send you the jar version of the binding ?

You could zip the jar and post it here on this issue for @paphko.

**Comment by svilenvul** [comment]
@paphko 
You can test this one: 
[org.eclipse.smarthome.binding.fsinternetradio_0.9.0.201608031807.jar.zip](https://github.com/eclipse/smarthome/files/399742/org.eclipse.smarthome.binding.fsinternetradio_0.9.0.201608031807.jar.zip)

**Comment by paphko** [comment]
Thank you @svilenvul, I'll test it asap. I need to setup OH2 first since I'm still using 1.8 :-)

**Comment by paphko** [comment]
This is the first time I'm using OH2 with Karaf and I cannot get your jar file running. If I simply put it into the addons folder, the binding is not found. After installing it via the Paper UI, another version (0.9.0-201608031951) is installed (listed via console bundle:list command).
@svilenvul: does this version also include your changes (at least the timestamp is a little bit later than yours)? Because that version works well.

**Comment by svilenvul** [comment]
@paphko - this version should not include my changes, because the PR is still not merged (I think so :))

I have downloaded the latest snapshot of the offline repo, unzipped the archive in the addons folder and started the runtime. The binding is listed with the bundle:list command and is in ACTIVE state. Maybe you can try restarting the runtime with system:shutdown and manually starting again.

**Comment by paphko** [comment]
@svilenvul I got your jar running and it works properly with my internet radio :-)

**Comment by kaikreuzer** [comment]
@paphko Thanks for testing, so we should be good to merge.
@svilenvul Could you rebase your branch as it currently shows merge conflicts? Thanks!

**Comment by svilenvul** [comment]
@kaikreuzer 
I have rebased the branch and squashed the commits into one. I guess this should be ready for a merge.

**Comment by kaikreuzer** [comment]
Thanks!

**Comment by maggu2810** [comment]
@svilenvul I assume you missed the Eclipse SmartHome launch configurations. They are broken now because of missing jetty client import contraint. You need to enable jetty client in the launch configuration.

**Comment by maggu2810** [comment]
I have created #1995

**Comment by watou** [comment]
@svilenvul @kaikreuzer Is it possible that [these reports](https://community.openhab.org/t/getting-a-npe-from-a-https-request/13610) are related to this PR?

**Comment by kaikreuzer** [comment]
Definitely yes. I created https://github.com/eclipse/smarthome/issues/2059 for it.
